Remove unnecessary methods from Store2, and moves handling of protobuf to `store`

jmuk opened the pull request:
**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
none
```

This introduces `Store2Backend`, and keep `pkg/config/crd` untyped (spec is `map[string]interface{}`), and move the conversion part to `proto.Message` into `pkg/config/store`. With this, it is easier to add other types of store like FSstore, the same conversion logic can be reused.

Also, we've agreed that the previous `Store2` interface had too many methods which wouldn't be used actually. This simplifies the interface, as removing `RegisterKind` (integrated into `Init`), removing kinds parameter from `Watch`, and removing of `Put` / `Delete`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/istio/mixer/1120)
<!-- Reviewable:end -->

istio-testing commented:
@jmuk: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/mixer-presubmit.sh | 15546ee3ef708613f928e58085c3322943e8cd5b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_mixer/1120/mixer-presubmit/318/) | `/test mixer-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@jmuk: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/mixer-presubmit.sh | 1beca9f4548a3981f970520ab8a82190bc166413 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_mixer/1120/mixer-presubmit/319/) | `/test mixer-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

mandarjog commented on pkg/config/store/queue.go:
```diff
@@ -12,58 +12,74 @@
 	for {
 		select {
 		case <-q.ctx.Done():
+			fmt.Printf("done0\n")
```
fmt? use glog.Info ?

jmuk replied:
oops, it was for my debugging. Removed.

istio-testing commented:
@jmuk: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/mixer-presubmit.sh | ace19effd88621ce10be47f55d47fbb73f23b077 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_mixer/1120/mixer-presubmit/320/) | `/test mixer-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

mandarjog commented on pkg/config/store/store2.go:
```diff
@@ -35,43 +36,96 @@ type Key struct {
+// Init initializes the connection with the storage backend. This uses "kinds"
+// for the mapping from the kind's name and its structure in protobuf.
+// The connection will be closed after ctx is done.
+func (s *Store2) Init(ctx context.Context, kinds map[string]proto.Message) error {
```
Please make Store2 and interface.

mandarjog commented on pkg/config/store/convert.go:
```diff
@@ -0,0 +1,46 @@
+
+func convert(spec map[string]interface{}, target proto.Message) error {
+	// This is inefficient; convert to a protobuf message through JSON.
+	// TODO: use reflect.
```
remove comment.
reflect will not work, it will not let us use ENUM symbolic constants.

jmuk replied:
done

istio-testing commented:
@jmuk: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/mixer-presubmit.sh | c9dc00293e84085b61de6a7217b7d482085122bf | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_mixer/1120/mixer-presubmit/322/) | `/test mixer-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

mandarjog commented on pkg/config/store/store2.go:
```diff
@@ -35,43 +36,110 @@ type Key struct {
+	backend Store2Backend
+}
+
+// NewStore2 creates a new Store2 instance with the specified backend.
```
We also need an abstract constructor that takes  `NewStore2(storeURL)` ?

jmuk replied:
Done

istio-testing commented:
@jmuk: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/mixer-presubmit.sh | 933df1e95fb0c687090c549c1b51e89aca69de14 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_mixer/1120/mixer-presubmit/326/) | `/test mixer-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

jmuk commented:
/retest

istio-testing commented:
@jmuk: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/mixer-presubmit.sh | 933df1e95fb0c687090c549c1b51e89aca69de14 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_mixer/1120/mixer-presubmit/328/) | `/test mixer-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

mandarjog commented on pkg/config/store/store2.go:
```diff
@@ -35,43 +38,143 @@ type Key struct {
 	Key
 	Type ChangeType
+
+	// Value refers the new value in the updated event. nil if the event type is delete.
```
The Value should be proto.Message, right ?

jmuk replied:
Done

istio-testing commented:
@jmuk: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/mixer-presubmit.sh | f96cd12321cb665e65dde1cf99d69a5ee6ef94b6 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_mixer/1120/mixer-presubmit/336/) | `/test mixer-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

jmuk commented:
/retest

istio-testing commented:
@jmuk: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/mixer-e2e-smoketest.sh | 27dc65bcc46ba2f17e644c1cd67be2db52f853d2 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_mixer/1120/mixer-e2e-smoketest/202/) | `/test mixer-e2e-smoketest`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

mandarjog commented on pkg/config/crd/store_test.go:
```diff
@@ -16,243 +16,180 @@ package crd
+	client := &Store{
+		conf:                 &rest.Config{},
+		discoveryBuilder:     createFakeDiscovery,
+		listerWatcherBuilder: lw,
```
aah ok, so listWatcher is for test injection.

jmuk replied:
yes.

mandarjog commented on pkg/config/store/queue.go:
```diff
@@ -12,46 +12,65 @@
-		chout:  make(chan store.Event),
-		chin:   make(chan store.Event),
+		ctx:   ctx,
+		chout: make(chan Event),
```
why is the `out chan` blocking chan? no length specified.

jmuk replied:
added

mandarjog commented on pkg/config/store/store2.go:
```diff
@@ -31,47 +34,154 @@ type Key struct {
+		return err
+	}
+
+	return convert(obj, spec)
```
obj is `map[string]interface{}` right?

jmuk replied:
yes

mandarjog commented on pkg/config/store/store2.go:
```diff
@@ -31,47 +34,154 @@ type Key struct {
+	data := s.backend.List()
+	result := make(map[Key]proto.Message, len(data))
+	for k, spec := range data {
+		pbSpec, err := convertWithKind(spec, k.Kind, s.kinds)
```
since we have `s.kinds` having `convert` and `converWithKind` methods is unnecessary.
you could access the map in the call.

jmuk replied:
The main point here is to create a clone of the message from `s.kinds`, and then convert `spec` into the clone (otherwise `s.kinds` will be affected). Also, the same logic is used in `queue.go`.

Maybe this structure wasn't clear. removed `convertWithKind`, create `cloneMessage` utility instead.

mandarjog commented:
Over all the PR looks good, 2 things though.

1. Please add one paragraph of design notes
2. Please increase the code coverage?  store2.go, store.go and resource.go have especially low code coverage.

jmuk commented:
Done. Also added tests. I think this improves the coverage (I'm afraid that `store2.go`'s coverage won't be very high, since it still has some code path used to connect with real k8s servers).

Updated the PR description to describe the overview of this change.

mandarjog commented:
@sebastienvas code coverage numbers are not being updated. how do we fix it ?

jmuk commented:
Per offline discussion, a new file `init.go` was created in `pkg/config/crd` to scope the code that requires actual kubernetes services for initialization.

Here's the local result of the code coverage:
```
pkg/config/crd/init.go: 0%
pkg/config/crd/resource.go: 100%
pkg/config/crd/store.go: 91.4%
pkg/config/store/convert.go: 100%
pkg/config/store/queue.go: 92.6%
pkgconfig/store/store2.go: 100%
```

mandarjog commented:
/approve

mandarjog commented:
/lgtm

sebastienvas commented:
@mandarjog Which numbers ? the one from codecov or the one in pkg_check. cc @yutongz
